### DISCUSSION TITLE
KDL 2.0.0 support

### DISCUSSION THREAD

**Discussion by zkat** [YaLTeR/niri#887] [Ideas]
[KDL 2.0.0 has been released!](https://github.com/kdl-org/kdl/releases/tag/2.0.0)

Along with this new spec release, I've released [`kdl-rs@6.0.0`](https://github.com/kdl-org/kdl-rs/releases/tag/v6.0.0), which supports both KDL 1.0.0 and KDL 2.0.0, and is able to translate document text between both versions.

This means that Niri would not need to move away from `knuffel` to add 2.0 support: by pulling in `kdl-rs` and enabling the `v1-fallback` feature, you can basically do something like this:

```rust
let mut doc = kdl::parse(input)?;
doc.ensure_v1();

let config: NiriConfig = knuffel::parse("example.kdl", doc.to_string()).expect("This should always succeed because `ensure_v1()` should guarantee a valid v1 doc, if the initial parse succeeded");

...continue as usual.
```

Nothing else would have to change about Niri's code unless you want to do your own heuristics around error management/fallbacks.

`kdl-rs`'s 2.0 parser is able to collect many syntax errors in a single pass, and so can give richer error messages than either the 1.0 parser (for now), or knuffel's parser.

Always happy to answer questions! I hope this is useful to y'all!

**Comment by YaLTeR** [comment]
Oh, it's finally out! Congrats on the release.

> Along with this new spec release, I've released [kdl-rs@6.0.0](https://github.com/kdl-org/kdl-rs/releases/tag/v6.0.0), which supports both KDL 1.0.0 and KDL 2.0.0, and is able to translate document text between both versions.
>
> This means that Niri would not need to move away from knuffel to add 2.0 support: by pulling in kdl-rs and enabling the v1-fallback feature, you can basically do something like this:

This is very very cool. Guess we don't have to have a big backwards compatibility break after all?

I will absolutely try this out when I'm finished with the floating PR. I'll also need to see if I can make it so when the config is KDL v2, it doesn't accept a few deprecated features.

**Reply by zkat** [reply]
Yeah I think it makes sense for Niri to have its own fallback logic, using parse_v1() and parse_v2() and making its own decision about what to do with the result. For example, if you use the  fallback as-is, and most folks have v1 configs, they will never get *v1* errors if they make a v1-specific mistake. It’ll just pretend it was supposed to be v2 the entire time. Probably something to play with heuristically.

**Reply by YaLTeR** [reply]
Hmm, thanks for the tip, I'll need to play around with it.

**Reply by zkat** [reply]
update: 6.2.0 now has "smarter" heuristics about which errors to report.
